pyro/distributions/polya_gamma.py

```diff
@@ -0,0 +1,70 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import math
+
+import torch
+from torch.distributions import constraints
+
+from pyro.distributions.torch import Gamma
+from pyro.distributions.torch_distribution import TorchDistribution
+
+
+class TruncatedPolyaGamma(TorchDistribution):
+    """
+    This is a PolyaGamma(1, 0) distribution truncated to have finite support in
+    the interval (0, 2.5). See [1] for details. As a consequence of the truncation
+    the `log_prob` method is only accurate to about six decimal places. In
+    addition the provided sampler is a rough approximation that is only meant to
+    be used in contexts where sample accuracy is not important (e.g. in initialization).
+    Broadly, this implementation is only intended for usage in cases where good
+    approximations of the `log_prob` are sufficient, as is the case e.g. in HMC.
+
+    :param tensor prototype: A prototype tensor of arbitrary shape used to determine
+        the `dtype` and `device` returned by `sample` and `log_prob`.
+
+    References
+
+    [1] 'Bayesian inference for logistic models using Polya-Gamma latent variables'
+        Nicholas G. Polson, James G. Scott, Jesse Windle.
+    """
+    truncation_point = 2.5
+    num_log_prob_terms = 7
+    num_gamma_variates = 8
+    assert num_log_prob_terms % 2 == 1
+
+    arg_constraints = {}
+    support = constraints.interval(0.0, truncation_point)
+    has_rsample = False
+
+    def __init__(self, prototype, validate_args=None):
+        self.dtype = prototype.dtype
+        self.device = prototype.device
+        super(TruncatedPolyaGamma, self).__init__(batch_shape=(), event_shape=(), validate_args=validate_args)
+
+    def expand(self, batch_shape, _instance=None):
+        new = self._get_checked_instance(TruncatedPolyaGamma, _instance)
+        super(TruncatedPolyaGamma, new).__init__(batch_shape, self.event_shape, validate_args=False)
+        new._validate_args = self.__dict__.get("_validate_args")
+        new.dtype = self.dtype
+        new.device = self.device
+        return new
+
+    def sample(self, sample_shape=()):
+        denom = torch.arange(0.5, self.num_gamma_variates).pow(2.0)
+        ones = torch.ones(self.batch_shape + sample_shape + (self.num_gamma_variates,),
+                          dtype=self.dtype, device=self.device)
+        x = Gamma(ones, ones).sample()
```

Comment:
Noting that `Gamma(1,1) == Exponential(1)`, you could speed this up via something like
```py
shape = self._extended_shape(sample_shape)
ones = self.prototype.new_ones(())
x = Exponential(ones).sample(shape)
```

Reply:
done